ranger/ext/img_display.py

```diff
@@ -465,3 +464,205 @@ def _get_sizes(self):
+                          struct.pack('HHHH', 0, 0, 0, 0))
+        n_cols, n_rows, x_px_tot, y_px_tot = struct.unpack('HHHH', ret)
+        self.pix_row, self.pix_col = x_px_tot / n_rows, y_px_tot / n_cols
+        self.needs_late_init = False
+
+    def draw(self, path, start_x, start_y, width, height):
+        self.image_id += 1
+        # dictionary to store the command arguments for kitty
+        # a is the display command, with T going for immediate output
+        # i is the id entifier for the image
+        cmds = {'a': 'T', 'i': self.image_id}
+        # sys.stderr.write('{}-{}@{}x{}\t'.format(start_x, start_y, width, height))
+
+        # finish initialization if it is the first call
+        if self.needs_late_init:
+            self._late_init()
+        with warnings.catch_warnings(record=True):  # as warn:
+            warnings.simplefilter('ignore', self.backend.DecompressionBombWarning)
+            image = self.backend.open(path)
+            # TODO: find a way to send a message to the user that
+            # doesn't stop the image from displaying
+            # if warn:
+            #     raise ImageDisplayError(str(warn[-1].message))
+        box = (width * self.pix_row, height * self.pix_col)
+
+        if image.width > box[0] or image.height > box[1]:
+            scale = min(box[0] / image.width, box[1] / image.height)
+            image = image.resize((int(scale * image.width), int(scale * image.height)),
+                                 self.backend.LANCZOS)
+
+        # start_x += ((box[0] - image.width) // 2) // self.pix_row
+        # start_y += ((box[1] - image.height) // 2) // self.pix_col
+        if self.stream:
+            # encode the whole image as base64
+            # TODO: implement z compression
+            # to possibly increase resolution in sent image
+            if image.mode != 'RGB' and image.mode != 'RGBA':
+                image = image.convert('RGB')
+            # t: transmissium medium, 'd' for embedded
+            # f: size of a pixel fragment (8bytes per color)
+            # s, v: size of the image to recompose the flattened data
+            # c, r: size in cells of the viewbox
+            cmds.update({'t': 'd', 'f': len(image.getbands()) * 8,
+                         's': image.width, 'v': image.height, })
+            payload = base64.standard_b64encode(
+                bytearray().join(map(bytes, image.getdata())))
+        else:
+            # put the image in a temporary png file
+            # t: transmissium medium, 't' for temporary file (kitty will delete it for us)
+            # f: size of a pixel fragment (100 just mean that the file is png encoded,
+            #       the only format except raw RGB(A) bitmap that kitty understand)
+            # c, r: size in cells of the viewbox
+            cmds.update({'t': 't', 'f': 100, })
+            with NamedTemporaryFile(prefix='rgr_thumb_', suffix='.png', delete=False) as tmpf:
```

Comment:
Even PNGs get resized most of the time in my experience, so an in termediate file is generated in most cases. I do not think that the extra logic and code bloat deriving from it would justify the very few cases were a resize is not necessary.
It would maybe be worthy in space-saving optic to change a bit the behavior of the caching for the thumbnails. To put it bluntly, it doesn't work for now. The idea was to keep around those temporary files so only the first time visiting a big image the user would experience a slowdown, and then it would be smooth sailing for the rest of the session. On low end hardware this would be even more important. However, I am not currently checking if the filename already exist, so the calculations are repeated **and** the files stick around until ranger is closed. I will see if I can get it to work properly, and add some extra logic in the `clear` method to clean up old cached files if they start accumulating.
If it takes too much time with the next release approaching disabling the chaching is actually trivial, I can make `NamedTemporaryFile` actually delete the file at the end of the `with` block and move all the file sending logic inside said block.

Reply:
If we hold off on caching I'd add the removal at the end of `NamedTemporaryFile`. I didn't consider PNGs would also have to be processed anyway.

Reply:
Actually why not use the full "ranger_thumb_" prefix?